RealmSwift/Map.swift

```diff
@@ -0,0 +1,600 @@
+ that are reported by Realm notifications.
+ */
+@frozen public enum RealmMapChange<Collection: RealmKeyedCollection> {
+
+    /**
+     `.initial` indicates that the initial run of the query has completed (if
+     applicable), and the collection can now be used without performing any
+     blocking work.
+     */
+    case initial(Collection)
+
+    /**
+     `.update` indicates that a write transaction has been committed which
+     either changed which objects are in the collection, and/or modified one
+     or more of the objects in the collection.
+
+     All three of the change arrays are always sorted in ascending order.
+
+     - parameter deletions:     The keys in the previous version of the collection which were removed from this one.
+     - parameter insertions:    The keys in the new collection which were added in this version.
+     - parameter modifications: The keys of the objects in the new collection which were modified in this version.
+     */
+    case update(Collection, deletions: [Collection.Key], insertions: [Collection.Key], modifications: [Collection.Key])
+
+    /**
+     If an error occurs, notification blocks are called one time with a `.error`
+     result and an `NSError` containing details about the error. This can only
+     currently happen if opening the Realm on a background thread to calcuate
+     the change set fails. The callback will never be called again after it is
+     invoked with a .error value.
+     */
+    case error(Error)
+
+    static func fromObjc(value: Collection?, change: RLMDictionaryChange?, error: Error?) -> RealmMapChange {
+        if let error = error {
+            return .error(error)
+        }
+        if let change = change {
+            return .update(value!,
+                           deletions: change.deletions as! [Collection.Key],
+                           insertions: change.insertions as! [Collection.Key],
+                           modifications: change.modifications as! [Collection.Key])
+        }
+        return .initial(value!)
+    }
+}
+
+// MARK: - RealmKeyedCollection Conformance
+
+extension Map: RealmKeyedCollection { }
+
+// MARK: - MapIndex
+
+/// Container type which holds the offset of the element in the Map.
+public struct MapIndex {
+    /// The position of the element in the Map.
+    public var offset: UInt
+}
+
+// MARK: - SingleMapEntry
+
+/// Container for holding a single key-value entry in a Map. This is used where a tuple cannot be expressed as a generic arguement.
+public struct SingleMapEntry<Key: _MapKey, Value: RealmCollectionValue>: _RealmMapValue, Hashable {
+    /// :nodoc:
+    public static func == (lhs: SingleMapEntry, rhs: SingleMapEntry) -> Bool {
+        return lhs.value == rhs.value
+    }
+    /// :nodoc:
+    public func hash(into hasher: inout Hasher) {
+        hasher.combine(key)
+    }
+    /// :nodoc:
+    public var key: Self.Key
+    /// :nodoc:
+    public var value: Self.Value
```

Comment:
These don't need very interesting docs but we do want these properties to appear in the generated docs.